File: 3720.lexicographically-smallest-permutation-greater-than-target.go
package leetcode

// @leet start

func lexGreaterPermutation(s string, target string) string {
	// N = len(str), N = len(target)
	// Case 1: k == M < N
	// prefix = target, fill remaining
	// Case 2: need to diverge before
	// backtrack
	const MAX_ALPHABETS = 26

	n := len(s)
	m := len(target)

	freqs := make([]int, MAX_ALPHABETS)
	for _, c := range s {
		freqs[toInt(c)]++
	}

	ans := make([]rune, 0, n)
	k := 0

	fillRemaining := func(str []rune) string {
		for i := range freqs {
			for freqs[i] > 0 {
				str = append(str, toRune(i))
				freqs[i]--
			}
		}
		return string(str)
	}

	greaterThan := func(c rune) int {
		idx := toInt(c)
		for i := idx + 1; i < MAX_ALPHABETS; i++ {
			if freqs[i] > 0 {
				return i
			}
		}
		return -1
	}

	// form prefix
	for i := range min(n, m) {
		c := rune(target[i])
		if freqs[toInt(c)] > 0 {
			freqs[toInt(c)]--
			ans = append(ans, c)
			k++
		} else {
			break
		}
	}

	// Case 1
	if k == m && n < m {
		return fillRemaining(ans)
	}

	// S == T, we put last char back under consideration
	if k == n {
		lastChar := ans[len(ans)-1]
		ans = ans[:len(ans)-1]
		freqs[toInt(lastChar)]++
		k--
		// if S == T, we must turn k back into the next char under consideration
		// 0 1 2 3 _    but _ is out-of-idx
		// a b c d
		// if k = 4, but S == T, we decr k to 3

	}

	// if prefixLen = 3, 3 is also the idx of the next char under consideration
	// 0 1 2 3
	//       ^   this is now the char under consideration
	// a b c _   ans
	//           we try to find a greater than _

	// Case 2
	// we already know backtracking is required immediately
	// k is the index of the char under consideration
	// i.e. for 0123_
	for i := k; i >= 0; i-- {
		nextGreater := greaterThan(rune(target[i]))
		if nextGreater >= 0 {
			ans = append(ans, toRune(nextGreater))
			freqs[nextGreater]--
			return fillRemaining(ans)
		}

		if i > 0 {
			lastChar := ans[len(ans)-1]
			ans = ans[:len(ans)-1]
			freqs[toInt(lastChar)]++
		}
	}

	return ""
}

func toRune(s int) rune {
	return rune(s + 'a')
}

func toInt(c rune) int {
	return int(c) - int('a')
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// @leet end

File: 3720.lexicographically-smallest-permutation-greater-than-target.cpp
// @leet imports start
#include "debug.hpp"
using namespace std;
// @leet imports end

// @leet start

class Solution {
  private:
    static constexpr int MAX_ALPHABETS = 26;
    constexpr char to_char(int x) { return static_cast<char>(x + 'a'); }
    constexpr int to_int(char c) { return c - 'a'; }

  public:
    string lexGreaterPermutation(string s, string target) {
        // using arbitrarily lengths of S and T
        // where |S| = n and |T| = m and |Sigma| = 26
        // taking n, m <= 2 * 10^5
        // hence necessitating TC of O( (n+m) * |sigma|)
        // with bitmask, can drop to O( n + m + |sigma|)

        // Case 1: k = m < n
        // i.e. T is a proper prefix of S
        // fill remaining alphabetically

        // Case 2a: k = n = m
        // i.e. S == T
        // no more chars to take from s

        // Case 2b: k = n < m
        // p is a proper prefix of T
        // no more chars to take from s

        // Case 3: k < n, m
        // i.e. divergence at k < min(n, m)

        std::string ans;
        ans.reserve(s.size());

        int n = s.size();
        int m = target.size();

        unsigned int mask = 0;
        vector<int> freqs(MAX_ALPHABETS);

        auto add_freq = [&](int k) {
            if (++freqs[k] >= 1) {
                mask |= 1 << k;
            }
        };

        auto lower_freq = [&](int k) {
            if (--freqs[k] == 0) {
                mask &= ~(1 << k);
            }
        };

        auto fill_remaining = [&](string& str) {
            for (int i = 0; i < MAX_ALPHABETS; ++i) {
                while (freqs[i] > 0) {
                    lower_freq(i);
                    str += to_char(i);
                }
            }
        };

        auto greater_than = [&](char c) {
            int x = to_int(c);
            // for (int i = x + 1; i < MAX_ALPHABETS; ++i) {
            //     if (freqs[i] > 0) {
            //         return i;
            //     }
            // }
            // return -1;

            // zero out bits <= current idx
            auto available = mask & (~0u << (x + 1));

            if (available == 0) {
                return -1;
            }
            // no. of consec 0s from lsb gives idx of next greater
            return std::countr_zero(available);
        };

        for (char c : s) {
            add_freq(to_int(c));
        }

        // Form prefix
        for (int i = 0; i < min(n, m); ++i) {
            if (freqs[to_int(target[i])] > 0) {
                ans += target[i];
                lower_freq(to_int(target[i]));
            } else {
                break;
            }
        }

        int k = ans.size();

        if (s.empty()) {
            return {};
        } else if (target.empty()) {
            fill_remaining(ans);
            return ans;
        }

        // Case 1, k == m < n
        if (k == m && n > m) {
            fill_remaining(ans);
            return ans;
        }

        // Case 2a
        // k == n == m
        // if S == T,
        // we must turn k back into the next char under consideration
        // since we can't consider idx[n]
        // e.g. if k == 4 == N <= M
        // 0 1 2 3 _
        // a b c d
        // there is no more n to be considered

        // Case 2b:
        // k == n < m
        // k has no more chars
        if (k == n) {
            add_freq(to_int(ans.back()));
            ans.pop_back();
            --k;
        }

        // Case 2+3
        // backtrack
        // k < n, m
        // k is the idx of char under consideration
        for (int i = k; i >= 0; --i) {
            int next_greater = greater_than(target[i]);
            if (next_greater != -1) {
                ans += to_char(next_greater);
                lower_freq(next_greater);
                fill_remaining(ans);
                return ans;
            }

            if (i > 0) {
                add_freq(to_int(ans.back()));
                ans.pop_back();
            }
        }
        return {};
    }
};
// @leet end

int main() {
    Solution s;
    string ss = "zaaz";
    string target = "az";
    auto out = s.lexGreaterPermutation(ss, target);
    debug(out);
}
